### PR TITLE
Add a check for circular references in container

### PR DESCRIPTION
**Pull request by ibrasho**
In response to https://github.com/laravel/framework/issues/7554

### CODE REVIEW

**Comment by ibrasho** [comment]
Just to explain:
This uses the `Container::$buildStack`. Whenever it's building something, it ensures that the `$buildStack` doesn't contains the concrete class it wants to instantiate.

It'll will throw a `BindingResolutionException` withe the first class and the current class.

Assume class A depends on class B which depends on class A again `A > B > A`.

I don't like the error message, but basically the first class name is the root of the dependencies `A` and the second class name is where the cyclic reference occurs which is `B`.

File: src/Illuminate/Container/Container.php [hunk 1]
```diff
@@ -785,6 +785,14 @@ public function build($concrete, $parameters = [])
 			throw new BindingResolutionException($message);
 		}
 
+		if (in_array($concrete, $this->buildStack))
+		{
+			$dependency = reset($this->buildStack);
+			$message = "Cyclic reference found while resolving [$concrete] and [$dependency].";
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/Container.php]: Don't define this, just dump it directly into the constructor.

File: tests/Container/ContainerTest.php [hunk 2]
```diff
@@ -596,3 +605,15 @@ function containerTestInject(ContainerConcreteStub $stub, $default = 'taylor')
 {
 	return func_get_args();
 }
+
+class ContainerCyclicReferenceStubA
+{
+	public function __construct(ContainerCyclicReferenceStubA $a)
+	{}
```
Comment (GrahamCampbell) [inline tests/Container/ContainerTest.php]: Just put these braces on the same line as the function definition.

File: src/Illuminate/Container/Container.php (hunk 1 above)
Comment (ibrasho) [inline src/Illuminate/Container/Container.php]: Check a few lines before this, the same thing is done.

Also if the string is created in the constructor, the line length is 123.

File: tests/Container/ContainerTest.php (hunk 2 above)
Comment (ibrasho) [inline tests/Container/ContainerTest.php]: Done.

**Comment by ibrasho** [comment]
I've changed the exception message since it was not really helpful.

Any suggestion for a better message?

I still a check for debug mode to added, since you most likely don't want this always to be turned on.

File: src/Illuminate/Container/Container.php [hunk 3]
```diff
@@ -785,6 +785,14 @@ public function build($concrete, $parameters = [])
 			throw new BindingResolutionException($message);
 		}
 
+		if (in_array($concrete, $this->buildStack))
+		{
+			$class = reset($this->buildStack);
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/Container.php]: Surely `$class = $this->buildStack[0]` is more efficient?

**Comment by ibrasho** [comment]
I've updated the commit, it feels much more better now. And exception message should have a clear meaning now.

File: src/Illuminate/Container/Container.php [hunk 4]
```diff
@@ -785,6 +785,13 @@ public function build($concrete, $parameters = [])
 			throw new BindingResolutionException($message);
 		}
 
+		if (in_array($concrete, $this->buildStack))
+		{
+			$message = "Cyclic reference found while resolving [$concrete].";
+
+			throw new BindingResolutionException($message);
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/Container.php]:
```php
throw new BindingResolutionException(
    "Cyclic reference found while resolving [$concrete]."
);
```

Comment (GrahamCampbell) [inline src/Illuminate/Container/Container.php]: Or just one line it provided it doesn't go over 120 chars.

Comment (GrahamCampbell) [inline src/Illuminate/Container/Container.php]: Also, I don't think `BindingResolutionException` is really what we want here. We can resolve the binding just fine. We need another exception for this case.

File: tests/Container/ContainerTest.php [hunk 5]
```diff
@@ -354,6 +354,30 @@ public function testInternalClassWithDefaultParameters()
...
+		$container->make('ContainerCyclicReferenceStubA', $parameters);
+	}
+
+
+	public function testCyclicReferenceCheckDetectCycleStartLocation()
+	{
+		// Since the dependency is ( D -> F -> E -> D ), the exception
+		// message should state that the issue starts in class D
+		$this->setExpectedException('Illuminate\Container\BindingResolutionException', "Cyclic reference found while resolving [ContainerCyclicReferenceStubD].");
+		$container  = new Container;
```
Comment (GrahamCampbell) [inline tests/Container/ContainerTest.php]: Don't align `=` please.

File: tests/Container/ContainerTest.php [hunk 6]
```diff
@@ -354,6 +354,30 @@ public function testInternalClassWithDefaultParameters()
...
+		$parameters = array();
```
Comment (GrahamCampbell) [inline tests/Container/ContainerTest.php]: Use short array syntax on all new code.

Comment (ibrasho) [inline tests/Container/ContainerTest.php]: Should I do it for all occurrences in the test file also?

Comment (GrahamCampbell) [inline tests/Container/ContainerTest.php]: No.

File: src/Illuminate/Container/Container.php (hunk 4 above)
Comment (ibrasho) [inline src/Illuminate/Container/Container.php]: I created a CircularReferenceException class. But I had to update the `Container::resolveClass()` method to catch the new exception.

File: tests/Container/ContainerTest.php (hunk 6 above)
Comment (GrahamCampbell) [inline tests/Container/ContainerTest.php]: Just **new** code, like I said.

Comment (ibrasho) [inline tests/Container/ContainerTest.php]: Done.

**Comment by ibrasho** [comment]
I switched the content of the test, since the new order makes more since in regard to the test names.

**Comment by ibrasho** [comment]
I still don't like the exception message. I would like to provide the full stack of dependencies that shows the circular reference.

**Comment by GrahamCampbell** [comment]
> I would like to provide the full stack of dependencies that shows the circular reference.

That's easy to do. Just pass the build stack to the exception and have a public getter.

**Comment by ibrasho** [comment]
Is this suitable? 
`Circular reference found while resolving [ContainerCircularReferenceStubB]. Dependency path: [ContainerCircularReferenceStubA -> ContainerCircularReferenceStubB -> ContainerCircularReferenceStubC -> ContainerCircularReferenceStubB].`

I tried add a `$buildStack` attribute to the exception, but I found that it does not make things easier. You will need to handle the specific exception to be able to get the build stack.

**Comment by GrahamCampbell** [comment]
> Is this suitable? 

No. Just do what I suggested. It's a pretty standard thing to do.

**Comment by ibrasho** [comment]
I just updated the PR.
Any more comments?

File: src/Illuminate/Container/CircularReferenceException.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+	/**
+	 * @param string $class
+	 * @param array $buildStack
+	 *
+	 * @return static
+	 */
+	public static function in($class, $buildStack)
+	{
+		return with(new static("Circular reference found while resolving [$class]."))->setBuildStack($buildStack);
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/CircularReferenceException.php]: Don't use "with" please.

File: src/Illuminate/Container/CircularReferenceException.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+	 */
+	protected $buildStack;
+
+	/**
+	 * @param string $class
+	 * @param array $buildStack
+	 *
+	 * @return static
+	 */
+	public static function in($class, $buildStack)
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/CircularReferenceException.php]: Just make this the constructor please.

File: src/Illuminate/Container/CircularReferenceException.php [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+use Exception;
+
+class CircularReferenceException extends Exception {
+
+	/**
+	 * @var array
+	 */
+	protected $buildStack;
+
+	/**
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/CircularReferenceException.php]: Need a "short description".

File: src/Illuminate/Container/CircularReferenceException.php [hunk 10]
```diff
@@ -0,0 +1,41 @@
+<?php namespace Illuminate\Container;
+
+use Exception;
+
+class CircularReferenceException extends Exception {
+
+	/**
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/CircularReferenceException.php]: Need a "short description".

File: src/Illuminate/Container/CircularReferenceException.php [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+	 * @param array $buildStack
+	 *
+	 * @return static
+	 */
+	public static function in($class, $buildStack)
+	{
+		return with(new static("Circular reference found while resolving [$class]."))->setBuildStack($buildStack);
+	}
+
+	/**
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/CircularReferenceException.php]: Need a "short description".

File: src/Illuminate/Container/CircularReferenceException.php [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+	 * @param $buildStack
+	 */
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/CircularReferenceException.php]: Missing a return annotation.

File: src/Illuminate/Container/CircularReferenceException.php [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+	public function setBuildStack($buildStack)
+	{
+		$this->buildStack = $buildStack;
+
+		return $this;
+	}
+
+	/**
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/CircularReferenceException.php]: Need a "short description".

**Comment by ibrasho** [comment]
I don't feel good about changing the default constructor. :-(

How does this look now?

File: src/Illuminate/Container/CircularReferenceException.php [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+class CircularReferenceException extends Exception {
+
+	/**
+	 * @var array
+	 */
+	protected $buildStack;
+
+    /**
+     *  Create a new Circular Reference Exception instance
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/CircularReferenceException.php]: Indent with tabs please, and also we need one space before the "C", here. ;)

Comment (GrahamCampbell) [inline src/Illuminate/Container/CircularReferenceException.php]: And a full stop, and all lower case apart from the first letter in this sentence.

Comment (ibrasho) [inline src/Illuminate/Container/CircularReferenceException.php]: When are we going to finally follow PSR-2... :mask: 

About the description, it was supposed to be CircularReferenceException. Should I change this?

Comment (GrahamCampbell) [inline src/Illuminate/Container/CircularReferenceException.php]: No, just make it "circular reference exception", just as you'd read it. It's a description. It's totally pointless to just repeat the class name. People already know what the class name is.

**Comment by taylorotwell** [comment]
Gonna hold off on this for now.

**Comment by GrahamCampbell** [comment]
:(

**Comment by theofidry** [comment]
Any update on this? It feels wrong to rely on the maximum function nesting level instead of detecting it early to throw a regular exception. Especially as services are lazy instantiated, it might not detect the issue before requiring this service. Even if you add tests to prevent that, I just get a `Segmentation fault` message, requiring check the services one by one to figure out which one is causing this. A simple `CircularReferenceException` exception with the service ID would solve this.

**Comment by balping** [comment]
+1 For start thinking about this again

**Comment by renepardon** [comment]
Would be nice to see where this kind of error comes from. Problem I have right now: calling an Artisan command (lighthouse) to generate new classes. Here I reach 256. Other framework commands work just fine. even xDebug Profiler doesn't seem to help much.

**Comment by macropay-solutions** [comment]
We found a [solution](https://github.com/macropay-solutions/maravel-framework/pull/35) for this.

**Pull request closed without merging**